### PR TITLE
Disable fixing files if fixing can cause bugs

### PR DESCRIPTION
**Pull request by bartlangelaan**
> Which issue, if any, is this issue related to?

Currently, it's related to the pinned issues #4119 and #2643 - but more issues can be added with the same logic.

> Is there anything in the PR that needs further explanation?

Yes.

Stylelint is an awesome tool, and I want to configure it for all projects we use at work. Because of the fixer, this would create almost no extra work for the team, because it can be run before somebody makes a commit.

There are however a few bugs open that make the autofixer behave unpredicted. I think the most known ones are #4119 and #2643. Looking at the comments in these issues, it looks like finding a solution for these issues is not easy.

I propose disabling the fixer when it detects that one of these issues may change the code in an unpredicted way. This way, the fixer doesn't always work, but it also never changes the code in an unpredicted way.

I added some code that tries to detect these bugs, and then disables the autofixer. I am unfamiliar with the code base of Stylelint, so bear with me 😄

There are probably better ways of coding this - if someone can can give me clues on what the 'right way' is, I'm happy to change my MR.

### CODE REVIEW

**Comment by jeddy3** [comment]
@bartlangelaan Thank you! I like this proposal a lot. As you said, it makes stylelint predictable and useable while we work to address the underlying issues, which are non-trivial.

We are aware that #4119 and #2643 are pressing issues, and we've started to drum up support for fixing them. So far, we've:

- relabeled them as ["high priority"](https://github.com/stylelint/stylelint/issues?q=is%3Aopen+is%3Aissue+label%3A%22priority%3A+high%22) and pinned them
- created [a pull request](https://github.com/stylelint/stylelint/pull/4561) to changes the docs to [draw more attention to the syntax parsers](https://github.com/stylelint/stylelint/blob/0bf44f76de65563fbd19f4f31c766fe07e909a5d/docs/about/syntaxes.md)

This week I plan to:

- fork the unmaintained parsers into the stylelint organisation
- create an umbrella issue for those issues associated with this parser

(See https://github.com/stylelint/stylelint/issues/4560#issuecomment-578907499)

I hope that the umbrella issue will become something the community rallies around as we need help supporting the unmaintained CSS-in-JS parser.

Neither fixing parser issues nor the disable comments one will be an easy task. For example, 
the proposal in https://github.com/stylelint/stylelint/issues/2643#issuecomment-535406358 to fix disable comments sounds good to me, but I suspect it'll be a breaking change that will require the adapting of all 81 autofixable rules. We should do this, but it may take some time. That is why I like your pull request as it improves stylelint in the interim. 

It would be great to see this proposal grow to address issues like https://github.com/stylelint/stylelint/issues/4560, but as you said #4119 and #2643 seem to be most common. What does everyone think about trialling this proposal as-is in the next patch release and then incorporating more issues afterwards?

File: lib/lintSource.js [hunk 1]
```diff
@@ -209,3 +215,54 @@ function createEmptyPostcssResult(filePath) {
...
+	// Check for issue #4119
+	if (root && root.source && root.source.lang === 'jsx' && root.nodes) {
+		const startEnds = root.nodes.map((n) =>
+			!n.source || !n.source.start || !n.source.input.css
+				? null
+				: {
+						startLine: n.source.start.line,
+						endLine: n.source.start.line + n.source.input.css.split('\n').length,
+				  },
+		);
```
Comment (hudochenkov) [inline lib/lintSource.js]: What do you think if we replace it with:

```js
		const startEnds = root.nodes.map((n) => {
			if (n.source && n.source.start && n.source.input.css !== undefined) {
				return {
					startLine: n.source.start.line,
					endLine: n.source.start.line + n.source.input.css.split('\n').length,
				};
			}

			return null;
		});
```

I think it's more understandable.

**Comment by jeddy3** [comment]
Once this goes in, we'll need to create a follow-up pull request to https://github.com/stylelint/stylelint/pull/4561 that document this approach within the:

- [about syntaxes section](https://github.com/stylelint/stylelint/blob/855fa8a4b92e16e42d7a896b28e84537d9d87ff3/docs/about/syntaxes.md)
- [usage options section](https://github.com/stylelint/stylelint/blob/855fa8a4b92e16e42d7a896b28e84537d9d87ff3/docs/user-guide/usage/options.md#fix)

The former section also seems like the best place to list the _syntax_ patterns that cause the autofixer to ignore file by this workaround.

I can make the follow-up pull request once we merge this and https://github.com/stylelint/stylelint/pull/4561.

---

I also suggest we don't close any issues. We should add a comment to an issue when we add a workaround and state that the underlying problem still needs fixing instead.

**Comment by bartlangelaan** [comment]
Hey @hudochenkov, I've rewritten the code that was unclear. I hope it's clearer now, please let me know if it's not.

File: lib/lintSource.js (hunk 1 above)
Comment (bartlangelaan) [reply lib/lintSource.js]: Done! I'm now pushing the objects into an array, so the filters to check for null values are not needed anymore.

File: lib/lintSource.js [hunk 2]
```diff
@@ -209,3 +215,54 @@ function createEmptyPostcssResult(filePath) {
...
+									n1.endLine < n2.startLine ||
+									// [b] [a]
+									n2.endLine < n1.startLine
+								)
+							),
+					),
+			)
+		) {
+			return false;
+		}
```
[the remark this answers is not in the document]
Comment (bartlangelaan) [reply lib/lintSource.js]: Rewritten this using for loops, and added comments so it's clear what every check is doing.

**Comment by hudochenkov** [comment]
> I also suggest we don't close any issues. We should add a comment to an issue when we add a workaround and state that the underlying problem still needs fixing instead.

Agree.

**Comment by jeddy3** [comment]
Changelog message added:

> -   Fixed: autofix will ignore sources containing disable comments or nested tagged template literals - this is workaround to make autofix safer to use until we can resolve the [underlying](https://github.com/stylelint/stylelint/issues/4119) [issues](https://github.com/stylelint/stylelint/issues/2643) ([#4573](https://github.com/stylelint/stylelint/pull/4573)).

**Comment by jeddy3** [comment]
Released in `13.2.0`.

**Pull request merged**